### PR TITLE
add mutation observer to rerender the components when we find a new element

### PR DESCRIPTION
**Pull request by rileylnapier**
## Description

drone deploy wants the inbox on different parts of different pages.  using a mutation observer will allow us to automatically rerender the inbox if the script has been initialized globally.

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> `Fix [#1]()`

### CODE REVIEW

File: packages/components/src/components/CourierSdk.tsx [hunk 1]
```diff
@@ -0,0 +1,47 @@
+import React, { useEffect } from "react";
+import { useCourier } from "@trycourier/react-provider";
+
+export const CourierSdk: React.FunctionComponent<{
```
Comment (suhasdeshpande) [inline packages/components/src/components/CourierSdk.tsx:4]: Is this intentional to suffice this with Sdk

File: packages/components/src/components/CourierSdk.tsx [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  activeComponents: {
+    inbox: boolean;
```
Comment (suhasdeshpande) [inline packages/components/src/components/CourierSdk.tsx:6]: maybe add some boolean-looking props? `isInboxActive`

File: packages/components/src/components/CourierSdk.tsx [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    toast: boolean;
+  };
+}> = ({ activeComponents, children }) => {
+  const courier = useCourier();
+
+  if (!window.courier.transport) {
```
Comment (suhasdeshpande) [inline packages/components/src/components/CourierSdk.tsx:12]: why not wrap this in an effect maybe?

File: packages/components/src/components/index.tsx [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+  const initialToast = document.querySelector("courier-toast");
+  const [toastElement, setToastElement] = useState(initialToast ?? undefined);
+
+  const toastConfig = {
+    ...componentConfigs?.toast,
+    ...getAttrsAsJson(toastElement),
+  };
+
+  useEffect(() => {
+    const observer = new MutationObserver((mutationsList) => {
```
Comment (suhasdeshpande) [inline packages/components/src/components/index.tsx:28]: can we create a function with a name for this to understand what this is supposed to be doing?

File: packages/components/src/index.html [hunk 5]
```diff
@@ -33,20 +33,34 @@
             document.body.appendChild(s);
           })();
         });
+      };*/
+      window.courierConfig = {
+        clientKey: "ZDRiMTViNmMtODdmZi00OWMzLWI5YTgtYmU5NTc3YTRlM2Y1",
+        userId,
+        apiUrl: "https://rubmz24skk.execute-api.us-east-1.amazonaws.com/dev",
```
Comment (suhasdeshpande) [inline packages/components/src/index.html:40]: does it have to be this particular URL? can it be api.courier.com instead?

File: packages/components/src/index.html [hunk 6]
```diff
@@ -33,20 +33,34 @@
...
+        wsUrl: "wss://20en15n3ng.execute-api.us-east-1.amazonaws.com/dev",
```
Comment (suhasdeshpande) [inline packages/components/src/index.html:41]: same as above. this feels bit leaky exposing gateway urls

File: packages/components/src/index.html [hunk 7]
```diff
@@ -33,20 +33,34 @@
...
+
+      window.addInbox = () => {
+        const existingComponent = document.querySelector("courier-inbox");
+        if (existingComponent) {
+          existingComponent.remove();
+
+          setTimeout(() => {
+            const inbox = document.createElement("courier-inbox");
+            document.body.appendChild(inbox);
+          }, 2000);
```
Comment (suhasdeshpande) [inline packages/components/src/index.html:52]: why 2000? would this break if we supply 0?

File: packages/components/src/components/index.tsx [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    observer.observe(document.body, {
+      attributes: false,
+      childList: true,
+      subtree: true,
+    });
+
+    return () => {
+      observer.disconnect();
+    };
+  }, [inboxElement, toastElement]);
```
Comment (rileylnapier) [inline packages/components/src/components/index.tsx:71]: remove these 2

**Review by suhasdeshpande — APPROVED** [review]

File: packages/components/src/components/CourierSdk.tsx [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+    window.courier.transport = courier.transport;
+  }
+
+  useEffect(() => {
+    for (const component of Object.keys(activeComponents)) {
+      const typedComponent = component as "inbox" | "toast";
+
+      if (!courier[typedComponent] || window.courier[typedComponent]) {
```
Comment (tk26) [inline packages/components/src/components/CourierSdk.tsx:20]: missing brackets around this before we negate?

Comment (tk26) [inline packages/components/src/components/CourierSdk.tsx:20]:
```suggestion
      if (!courier[typedComponent] || !window.courier[typedComponent]) {
```

Comment (tk26) [reply packages/components/src/components/CourierSdk.tsx:20]: not sure; do we continue or return if no courier component found?

**Pull request merged**